lib/pbench/server/api/resources/__init__.py

```diff
@@ -41,9 +42,38 @@ def __str__(self) -> str:
         return self.message
 
 
-class UnauthorizedAccess(APIAbort):
+class APIInternalError(APIAbort):
+    """Used to report a server internal error with a UUID value that connects
+    the string reported to the client with a server log entry to aid analysis
+    by an SRE
     """
-    The user is not authorized for the requested operation on the specified
+
+    def __init__(self, logger: Logger, message: str):
+        """Construct an "internal server error" exception object.
+
+        This exception is raised and will be used in the API _dispatch method
+        to generate a JSON error message payload to the client. This message
+        contains a UUID string that uniquely identifies this internal server
+        error context. An internal server log message is generated, with
+        traceback, also containing that UUID value so that a server
+        administrator can investigate the cause of the error.
+
+        NOTE: We want to return minimal explanations of internal errors to the
+        client while capturing detailed information for server developers to
+        determine what happened.
+
+        Args:
+            logger  Python logger
+            message A brief descriptive message for the client and log
+        """
+        self.uuid = uuid.uuid4()
+        msg = f"INTERNAL error {message!r}: report error code {self.uuid} to an SRE"
+        super().__init__(http_status=HTTPStatus.INTERNAL_SERVER_ERROR, message=msg)
+        logger.exception("{}", msg)
```

Comment:
> I suppose there is another alternative: if the message produced a "unique ID" _for the message_ as opposed to what we have now which is an ID for the _instance_, then we could look the ID up directly in the source code

Yeah, some components in OneView did this; it's a massive pain to manage, and I don't think anyone liked it.

Reply:
> Yeah, but I think Peter would lobby for something closer to "eh?" (like "Internal server error"...), and, given that level of paranoia

I prefer giving a bit more information without details, but I'll admit to being not entirely on one side or the other of this fence. I suspect one of my road bumps is the awkward use of "internal error" for partial completion; this at least loosely suggests retryability (which is usually what we want) and there's no actual failure that works. But that's an anomaly I hope to resolve when I can leverage transactions, so maybe I should stop worrying. Or maybe not.

Reply:
> some components in OneView did this; it's a massive pain to manage, and I don't think anyone liked it.

It strikes me as very handy, but also rather "costly" to implement and administer.  (In the VMS days, we maintained the messages separately from the status codes, so we had status codes for all the messages, and so it wasn't too hard to link the message back to the code...but, I don't think we have the infrastructure for that sort of thing on Linux.)

> I prefer giving a bit more information without details, but I'll admit to being not entirely on one side or the other of this fence.

What I want is to be able to find the relevant piece of code when the User sends me the text of the error message.  The UUID solution is sufficient (although it requires indirection through the log).  I would rather be able to go straight to the code, but, once I get there, I may want the other details that the log can supply.

I sympathize with the objection to keeping the user in the dark, as I myself will likely be that user at some point; however, in most use-cases, the user won't be able to do anything with the information (i.e., in the case of an internal error), and so a boilerplate message with the UUID should be fine.  (And, it keeps us from having to discuss whether the content of each message is generic enough without being useless.)

> I suspect one of my road bumps is the awkward use of "internal error" for partial completion; this at least loosely suggests retryability (which is usually what we want) and there's no actual failure that works.

I think, in that case, we should use some other error status, then.  There's _got_ to be one which we can bend to say, "check the payload for determination of success".

Apparently [`200` doesn't necessarily mean "success"](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/200):
> `PUT` or `POST`: The resource describing the result of the action is transmitted in the message body.

So, that would work for "partial success".

Reply:
> No, 200 means success

Yeah, in the sense that the server survived attempting the requested action and was able to send a response....  😆  

So, `200` means success in that the client received a response to the request; beyond that, it means the response body describes the result of the action...I'm not hearing any immediate indication of the success of the requested _operation_...it was only the request itself which was received and responded to successfully.

Anyway, I think that `200` closer to what we want than `500` is.

As for your responses for `PUT /datasets/metadata`, I would pull for something more like what the ES bulk ingestion does, although I think I would return a dictionary keyed by request item rather than a list of objects.  So,
```
{
    "metadata": {
        "global": {"dashboard": {"seen": True}
    },
    "errors": {
        "mine": "Key 'mine' value for 'global.mine.foo' is not a JSON object"
    }
}
```
Being able to iterate over `response["errors"].items()` and receiving the values of the erroneous keys without having to parse the messages seems like a good thing.

Reply:
> > No, 200 means success
> 
> Yeah, in the sense that the server survived attempting the requested action and was able to send a response.... laughing

So, you've just proven the the whole 4xx error space is pointless, because we should just return 200 and put error information in the response? 🙄 I don't buy that! 200 is a success. The standard doesn't make any provision for partial success, though, so we're fudging something whatever we do, and 500 was never more than a potential "least worst" response. I don't like 200, either; but if we drop actual details from 500 responses then that drops way down in the "worst" list and maybe 200 is the new least worst?

And, yeah, I started out with an object for errors and dropped it to a list for simplicity; but I will buy your argument here. (Though it'd be `"global.mine.foo": "error..."` since that's the key the client gave.)

Reply:
> you've just proven the the whole 4xx error space is pointless

No, the 4xx space is for requests which couldn't be executed (comments about the server's survival aside... 😀), whereas, in the case in point, the requested action could be performed but the outcome wasn't successful in all cases.

I concur that `200` is least worst (since you've ruled out `207`).  As success indicators go, it's way better than `202`:  "Um, I got your request, and...maybe it will get done."  🤣 

> it'd be `"global.mine.foo": "error..."` since that's the key the client gave.

Yeah, I couldn't be sure what the request was based solely on the response, but I guess I should have gone with that as a guess.  🍤   Still, I think it's better than a simple list.

Reply:
_[Been sitting on [that comment](https://github.com/distributed-system-analysis/pbench/pull/3091#discussion_r1043842286) for a while, eh, Dave?  😁  I think comments here are displayed in the order in which they were **created** (i.e., started), but timestamped in the order that they were **posted**....]_

> the client has no clue that it needs to look at the response JSON

Yep, that bothers me as well...which I why I keep coming back to `207`.  We're basically itching for a protocol violation of some sort here, so maybe returning JSON with that status is the least of evils.

Reply:
> _[Been sitting on [that comment](https://github.com/distributed-system-analysis/pbench/pull/3091#discussion_r1043842286) for a while, eh, Dave? grin I think comments here are displayed in the order in which they were **created** (i.e., started), but timestamped in the order that they were **posted**....]_
> 
> > the client has no clue that it needs to look at the response JSON
> 
> Yep, that bothers me as well...which I why I keep coming back to `207`. We're basically itching for a protocol violation of some sort here, so maybe returning JSON with that status is the least of evils.

I don't see any advantage in 207. Yeah, it's weird, which may or may not be significant, but it's also really obscure; and most anyone (including requests' `raise_from_status`) will correctly treat *any* 2xx as "success" and we'll just document the partial completion returns and, one way or another, clients will have to expect something non-standard.